File: programmers/소수 만들기/윤연상.py
def make_prime():
    tmp = [[i, True] for i in range(2, 3000)]
    
    for num, isPrime in tmp:
        if isPrime:
            for mul in range(num*2, 3000, num):
                tmp[mul-2][1] = False
    
    ret = []
    for num, isPrime in tmp:
        if isPrime:
            ret.append(num)
    return ret

def dfs(start, selected, nums, primes):
    if len(selected) == 3:
        if sum(selected) in primes:
            return 1
        else:
            return 0
    cnt = 0
    
    for i in range(start, len(nums)):
        selected.append(nums[i])
        cnt += dfs(i+1, selected, nums, primes)
        selected.pop()
    
    return cnt
    
def solution(nums):
    primes = make_prime()
    
    return dfs(0, [], nums, primes)

File: programmers/소수 만들기/윤연상.cpp
#include <vector>
#include <iostream>
using namespace std;

int ans = 0;
int picked[3];
bool isPrime[3001];

// 에레토스테네스의 체를 이용하여 소수를 구함
void find_prime_number() {
    for (int i = 2; i < 3000; i++)
        isPrime[i] = true;

    for (int i = 2; i < 3000; i++) {
        if (isPrime[i]) {
            for (int j = i * 2; j < 3000; j += i) {
                isPrime[j] = false;
            }
        }
    }
}
// dfs를 이용하여 3개의 수를 선택 후 해당 수가 소수인지 검사
void pick_number(int idx, int cnt, vector<int> nums) {
    if (cnt == 3) {
        int sum = picked[0] + picked[1] + picked[2];
        if (isPrime[sum]) 
            ans++;
        return;
    }

    for (int i = idx; i < nums.size(); i++) {
        picked[cnt] = nums[i];
        pick_number(i + 1, cnt + 1, nums);
    }
}

int solution(vector<int> nums) {
    find_prime_number();

    pick_number(0, 0, nums);

    return ans;
}
